### ISSUE TITLE
Memory leak - linkerd proxy 

### ISSUE THREAD

**Issue by gokcesurenkok** [linkerd/linkerd2#5734]
## Bug Report
Proxy injected pod is having a memory leak in time 

### What is the issue?
after the pod is started to run , linkerd proxy's memory consumption is just keep rising in time.
<img width="1696" alt="Screen Shot 2021-02-12 at 09 38 17" src="https://user-images.githubusercontent.com/9136196/107738850-961c1e80-6d18-11eb-8819-a2c9773abdd6.png">
### How can it be reproduced?
i have no idea how it can be reproduced. other services has no problem like this 
### Logs, error output, etc

https://gist.github.com/gokcesurenkok/5d1c410bcb4ab88d5fdd005c6d88bd85

#### `linkerd check` output

```
linkerd check
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

kubernetes-version
------------------
√ is running the minimum Kubernetes API version
√ is running the minimum kubectl version

linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
√ controller pod is running
√ can initialize the client
√ can query the control plane API

linkerd-config
--------------
√ control plane Namespace exists
√ control plane ClusterRoles exist
√ control plane ClusterRoleBindings exist
√ control plane ServiceAccounts exist
W0212 13:51:11.826795    2166 warnings.go:67] apiextensions.k8s.io/v1beta1 CustomResourceDefinition is deprecated in v1.16+, unavailable in v1.22+; use apiextensions.k8s.io/v1 CustomResourceDefinition
√ control plane CustomResourceDefinitions exist
W0212 13:51:11.832769    2166 warnings.go:67] admissionregistration.k8s.io/v1beta1 MutatingWebhookConfiguration is deprecated in v1.16+, unavailable in v1.22+; use admissionregistration.k8s.io/v1 MutatingWebhookConfiguration
√ control plane MutatingWebhookConfigurations exist
W0212 13:51:11.836748    2166 warnings.go:67] admissionregistration.k8s.io/v1beta1 ValidatingWebhookConfiguration is deprecated in v1.16+, unavailable in v1.22+; use admissionregistration.k8s.io/v1 ValidatingWebhookConfiguration
√ control plane ValidatingWebhookConfigurations exist
√ control plane PodSecurityPolicies exist

linkerd-identity
----------------
√ certificate config is valid
√ trust anchors are using supported crypto algorithm
√ trust anchors are within their validity period
√ trust anchors are valid for at least 60 days
√ issuer cert is using supported crypto algorithm
√ issuer cert is within its validity period
√ issuer cert is valid for at least 60 days
√ issuer cert is issued by the trust anchor

linkerd-webhooks-and-apisvc-tls
-------------------------------
√ tap API server has valid cert
√ tap API server cert is valid for at least 60 days
W0212 13:51:11.880217    2166 warnings.go:67] admissionregistration.k8s.io/v1beta1 MutatingWebhookConfiguration is deprecated in v1.16+, unavailable in v1.22+; use admissionregistration.k8s.io/v1 MutatingWebhookConfiguration
√ proxy-injector webhook has valid cert
√ proxy-injector cert is valid for at least 60 days
W0212 13:51:12.133479    2166 warnings.go:67] admissionregistration.k8s.io/v1beta1 ValidatingWebhookConfiguration is deprecated in v1.16+, unavailable in v1.22+; use admissionregistration.k8s.io/v1 ValidatingWebhookConfiguration
√ sp-validator webhook has valid cert
√ sp-validator cert is valid for at least 60 days

linkerd-api
-----------
√ control plane pods are ready
√ control plane self-check
√ [kubernetes] control plane can talk to Kubernetes
√ [prometheus] control plane can talk to Prometheus
√ tap api service is running

linkerd-version
---------------
√ can determine the latest version
√ cli is up-to-date

control-plane-version
---------------------
√ control plane is up-to-date
√ control plane and cli versions match
W0212 13:51:13.311694    2166 warnings.go:67] admissionregistration.k8s.io/v1beta1 MutatingWebhookConfiguration is deprecated in v1.16+, unavailable in v1.22+; use admissionregistration.k8s.io/v1 MutatingWebhookConfiguration

linkerd-prometheus
------------------
√ prometheus add-on service account exists
√ prometheus add-on config map exists
√ prometheus pod is running

linkerd-grafana
---------------
√ grafana add-on service account exists
√ grafana add-on config map exists
√ grafana pod is running

Status check results are √

```

### Environment

- Kubernetes Version: 1.19.7
- Cluster Environment: on prem
- Host OS: centos 7
- Linkerd version: stable-2.9.3

**Comment by kleimkuhler** [comment]
@gokcesurenkok Is there anything reproducible that you can provide for this? I know you mentioned you don't see it in other services, but for the service that you do see it on, can you provide any additional details about it?

An example manifest that reproduces it would be most helpful, but it would also be nice to know what kind of traffic this service handles.

**Comment by hawkw** [comment]
@gokcesurenkok Do you happen to have graphs of any other metrics from the effected proxy? It would be useful to see the values of the `tcp_open_connections`, `stack_create_total` and `stack_drop_total` metrics over time, as they might help determine what resource is being leaked and why.

**Comment by hawkw** [comment]
It would also be very useful to know what kind(s) of traffic are transiting the effected proxy, particularly in the outbound direction:

- Is the injected service making HTTP requests?
- Is it using non-HTTP network protocols? 
- Is the injected service encrypting its own outbound traffic, or is Linkerd responsible for TLS?

Similarly, what kind of incoming traffic is being received by this pod?

**Comment by gokcesurenkok** [comment]
hey @hawkw i attached linkerd metrics of one of the services pod 
https://gist.github.com/gokcesurenkok/24bf5fd1a9c62759e5854ca2eb88db88
and i am also attaching some screenshots if they help 
<img width="1727" alt="Screen Shot 2021-02-14 at 00 05 06" src="https://user-images.githubusercontent.com/9136196/107861720-a509ea80-6e58-11eb-9ffc-c715118b8388.png">
<img width="800" alt="Screen Shot 2021-02-14 at 00 03 48" src="https://user-images.githubusercontent.com/9136196/107861723-a804db00-6e58-11eb-844b-b9f397a65a6a.png">
<img width="918" alt="Screen Shot 2021-02-14 at 00 03 13" src="https://user-images.githubusercontent.com/9136196/107861726-aa673500-6e58-11eb-86ac-a5f8d38a5d92.png">
<img width="1697" alt="Screen Shot 2021-02-14 at 00 01 14" src="https://user-images.githubusercontent.com/9136196/107861727-acc98f00-6e58-11eb-8c28-238bb62c25a1.png">


and for your questions, it makes HTTP requests since there is a new relic agent in it, and it talks to other services by using grpc connections
and it is not encrypting its own connection, linkerd is responsible for that.

**Comment by hawkw** [comment]
Interesting, it seems like the `stack_create_total` and `stack_drop_total` metrics more or less line up, so we don't seem to be leaking stacks:
```
# HELP stack_create_total Total number of services created
# TYPE stack_create_total counter
stack_create_total{direction="outbound",name="balance.endpoint"} 358951
stack_create_total{direction="outbound",name="concrete"} 8974
stack_create_total{direction="outbound",name="source"} 11256
stack_create_total{direction="inbound",name="logical"} 168
stack_create_total{direction="inbound",name="source"} 564633
# HELP stack_drop_total Total number of services dropped
# TYPE stack_drop_total counter
stack_drop_total{direction="outbound",name="balance.endpoint"} 358866
stack_drop_total{direction="outbound",name="concrete"} 8971
stack_drop_total{direction="outbound",name="source"} 11250
stack_drop_total{direction="inbound",name="logical"} 164
stack_drop_total{direction="inbound",name="source"} 563781
```
and, the proxies in pods affected by the leak seem to have slightly _fewer_ file descriptors open than proxies in unaffected services (it looks like the `rohanapi` proxies exhibit the leak, while the `moriaapi` proxies don't).

**Comment by olix0r** [comment]
@gokcesurenkok Are you able to test the most recent edge release? Or, if you're uneasy about upgrading, perhaps you can try using a more recent proxy version by setting a pod spec annotation like `config.linkerd.io/proxy-version: edge-21.2.2`? We're starting to prepare the next stable release, and it would be helpful to determine whether this issue is still present. Much has changed with regards to caching and discovery since stable-2.9.

**Comment by gokcesurenkok** [comment]
hey @olix0r, luckily we have active-active two k8s clusters in production environment, so I was able to test the latest proxy version` edge-21.2.3` in one cluster. I am attaching the result.
<img width="1717" alt="Screen Shot 2021-02-19 at 21 50 54" src="https://user-images.githubusercontent.com/9136196/108550663-eb9b9100-72ff-11eb-99ab-229d03f1ad8f.png">

the left side is stable-2.9.3 and the right side is the latest edge version, it seems the problem has been solved. I will be upgrading the control plane and data planes accordingly to the edge version. Thanks for the support.
